### PR TITLE
Optimize tagged value eq when immediate

### PR DESCRIPTION
**Pull request by manuranga**
Part of #893

This pr doesn't handle the case where one side is a const

### CODE REVIEW

**Comment by manuranga** [comment]
```ballerina
public function f(1|"one"|2|"two"|boolean x, 2|"two"|3|"three"|boolean y) returns boolean {
    return x == y;
}
```
gets optimized to
```llvm
define i1 @_B04rootf(i8 addrspace(1)* readnone %0, i8 addrspace(1)* readnone %1) local_unnamed_addr !dbg !11 {
  %3 = alloca i8, align 1
  %4 = load i8*, i8** @_bal_stack_guard, align 8
  %5 = icmp ult i8* %3, %4
  br i1 %5, label %8, label %6

6:                                                ; preds = %2
  %7 = icmp eq i8 addrspace(1)* %0, %1
  ret i1 %7

8:                                                ; preds = %2
  %9 = call i8 addrspace(1)* @_bal_panic_construct(i64 2308), !dbg !26
  call void @_bal_panic(i8 addrspace(1)* %9)
  unreachable
}
```

File: compiler/modules/nback/equal.bal [hunk 1]
```diff
@@ -138,18 +135,53 @@ function buildEqualFloat(llvm:Builder builder, Scaffold scaffold, boolean exact,
...
-function isAnyOperandSmallString(bir:Operand[] operands) returns boolean {
-    foreach var operand in operands {
-        if operand is bir:StringConstOperand {
-            string str = operand.value;
-            byte[] bytes = str.toBytes();
-            int nBytes = bytes.length();
-            if isSmallString(str.length(), bytes, nBytes) {
-                return true;
+function isTypeImmediate(t:SemType ty) returns boolean {
+    if t:isSubtypeSimple(ty, t:NIL|t:BOOLEAN) {
```
Comment (jclark) [inline compiler/modules/nback/equal.bal:139]: Don't see the need for this.

File: compiler/modules/nback/equal.bal [hunk 2]
```diff
@@ -138,18 +135,53 @@ function buildEqualFloat(llvm:Builder builder, Scaffold scaffold, boolean exact,
...
+
+    // SUBSET extend when singleton subtypes of float and decimal are supported
+    if ty is t:UniformTypeBitSet || !t:isSubtypeSimple(ty, t:NIL|t:BOOLEAN|t:INT|t:STRING) {
+        return false;
+    }
+    return isIntSubtypeImmediate(ty) && isStringSubtypeImmediate(ty);
+}
+
+function isIntSubtypeImmediate(t:ComplexSemType ty) returns boolean {
+    t:IntSubtype|boolean intSubtype = t:intSubtype(ty);
```
Comment (jclark) [inline compiler/modules/nback/equal.bal:151]:
```
record {|
  int? min;
  int? max;
  boolean contiguous;
|};
```

**Review by jclark — CHANGES_REQUESTED** [review]

**Comment by manuranga** [comment]
Close in favor of #912.

**Pull request closed without merging**